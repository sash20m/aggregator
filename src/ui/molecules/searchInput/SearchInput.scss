.search-content {
    display: flex;
    width: 100%;
    flex-direction: column;
    align-items: center;

    &__input {
        width: 83vw;
        height: 8vh;
        border: 1px solid #9aabc2;
        color: rgba(0, 0, 0, 0.65);
        font-size: 1.3em;
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
        line-height: 1.5;
        border-radius: 50px;
        margin-top: 30px;
        padding-left: 30px;
        outline: none;

        &--active {
            width: 83vw;
            height: 8vh;
            border: 1px solid #467bbc;
            color: rgba(0, 0, 0, 0.65);
            font-size: 1.3em;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
            line-height: 1.5;
            margin-top: 30px;
            padding-left: 30px;
            outline: none;
            border-radius: 31px 31px 0 0;
        }
    }

    &__suggestions {
        position: absolute;
        z-index: 2;
        width: 83.15vw;
        overflow-y: auto;
        list-style: none;
        padding-left: 0;
        background: white;
        border-top-width: 0;
        border-bottom-left-radius: 20px;
        border-bottom-right-radius: 20px;
        border-left: 1px solid #467bbc;
        border-right: 1px solid #467bbc;
        border-bottom: 1px solid #467bbc;
        padding-right: 30px;
        margin-top: 5.75em;

        &__item {
            position: relative;
            overflow: hidden;
            font-size: 1em;
            font-weight: 600;
            line-height: 22px;
            color: #000;
            white-space: nowrap;
            text-overflow: ellipsis;
            cursor: pointer;
            box-sizing: border-box;
            list-style: none;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;

            &__button {
                width: 100%;
                text-align: left;
                padding: 10px 25px;
                border: none;
                margin: 0;
                overflow: visible;
                background: transparent;
                color: inherit;
                font: inherit;
                line-height: normal;
                -webkit-font-smoothing: inherit;
                -moz-osx-font-smoothing: inherit;
                -webkit-appearance: none;
                cursor: pointer;
                outline: inherit;

                &:hover {
                    color: #056ac8;
                }
            }
        }


    }

    &__options {
        display: flex;
        flex-direction: row;
        align-items: center;
        margin-top: 20px;

        @media only screen and (max-width: 300px) {
            flex-direction: column;

        }

        &:hover {
            cursor: pointer;
        }

        &__item {
            font-size: 1.1em;
            color: rgb(61, 61, 61);
            font-weight: 500;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
            margin: 5px;

            &--selected {
                font-size: 0.9em;
                color: #467bbd;
                font-weight: 500;
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
                margin: 5px;
                border-radius: 20px;
                border: 1px solid rgb(121, 121, 121);
                padding: 10px 15px 10px 15px;
                margin-left: 10px;
                margin-right: 10px;
            }
        }
    }
}