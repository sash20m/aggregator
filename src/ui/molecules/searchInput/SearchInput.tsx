/* eslint-disable jsx-a11y/mouse-events-have-key-events */
import React, { useEffect, useState } from 'react';

import { useRouter } from 'next/router';
import { getSuggestions } from 'services/companies.services';
import { Suggestions } from 'index';

import './SearchInput.scss';
import Link from 'next/link';

interface Props {
  suggestions: Suggestions[] | null;
}

export const SearchInput: React.FC<Props> = ({ suggestions }) => {
  const Router = useRouter();
  const [focused, setFocused] = useState<boolean>(false);
  const [mouseOver, setMouseOver] = useState<boolean>(false);
  const [data, setData] = useState<Suggestions[] | null>(null);
  const [inputValue, setInputValue] = useState('');
  const [lastSearched, setLastSearched] = useState<string | null>('');
  const [specialClass, setSpecialClass] = useState(false);

  const getSuggestionsList = async (name: string | null) => {
    try {
      const res = await getSuggestions(name);
      setData(res);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
    }
  };

  useEffect(() => {
    const onSetData = () => {
      setData(suggestions);
    };
    if (suggestions) onSetData();
    else {
      getSuggestionsList(null);
      setSpecialClass(true);
    }
  }, [suggestions]);

  useEffect(() => {
    if (Router) {
      if (Router.pathname.search('companies/')) {
        setLastSearched(sessionStorage.getItem('lastSearched'));
      }
    }
  }, [Router]);

  const searchCompanies = (e: React.ChangeEvent<HTMLInputElement>) => {
    getSuggestionsList(e.target.value);
    setInputValue(e.target.value);
  };

  const onInputClick = () => {
    setFocused(true);
  };

  const onInputLeave = () => {
    if (!mouseOver) setFocused(false);
  };

  const mouseOverList = () => {
    setMouseOver(true);
  };

  const mouseLeaveList = () => {
    setMouseOver(false);
  };

  const goToCompanies = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      sessionStorage.setItem('lastSearched', inputValue);
      Router.push(`/companies/${inputValue}`);
    }
  };

  const resetSearch = () => {
    sessionStorage.clear();
    Router.push(`/companies/all`);
  };

  return (
    <div className="search-content">
      <div className="search-content__input-wrapper">
        <input
          data-testid="search-input"
          className={
            focused ? 'search-content__input--active' : 'search-content__input'
          }
          type="text"
          placeholder="Search from 225,195 companies"
          onClick={onInputClick}
          onBlur={onInputLeave}
          onChange={searchCompanies}
          value={inputValue}
          onKeyDown={goToCompanies}
        />
        <img
          className="search-content__search-magnifier"
          src="../search.png"
          alt=""
        />
        {focused && (
          <ul
            data-testid="suggestion-items"
            className={
              specialClass
                ? 'search-content__special-suggestions'
                : 'search-content__suggestions'
            }
            onMouseOver={mouseOverList}
            onMouseOut={mouseLeaveList}
          >
            {!data ? (
              <p className="search-content__suggestions__loading">Loading...</p>
            ) : (
              data.map((company: Suggestions) => (
                <Link href="/company/[slug]" as={`/company/${company.slug}`}>
                  <li
                    key={company.name}
                    className="search-content__suggestions__item"
                  >
                    <button
                      type="button"
                      className="search-content__suggestions__item__button"
                      name={company.slug}
                      // onClick={goToCompany}
                    >
                      {`${company.name} â€¢ ${company.idno}`}
                    </button>
                  </li>
                </Link>
              ))
            )}
          </ul>
        )}
      </div>

      <div className="search-content__options">
        <p className="search-content__options__item">Search In</p>
        <p className="search-content__options__item--selected">Companies</p>
        <p className="search-content__options__item">Persons</p>
      </div>
      {lastSearched && (
        <div className="search-content__filter-wrapp">
          <div className="search-content__filter-wrapp__filter">
            <div className="search-content__filter-wrapp__filter__info">
              <img
                className="search-content__filter-wrapp__filter__icon"
                src="../filter.png"
                alt=""
              />
              <p>{lastSearched}</p>
            </div>

            <button
              className="search-content__filter-wrapp__filter__remove"
              type="button"
              onClick={resetSearch}
            >
              X
            </button>
          </div>
        </div>
      )}
    </div>
  );
};
