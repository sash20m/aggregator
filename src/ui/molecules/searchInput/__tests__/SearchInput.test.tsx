import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { SearchInput } from '../SearchInput';

afterEach(cleanup);

describe('searchInput', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<SearchInput suggestions={null} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should open the input', async () => {
    const { getByTestId } = render(<SearchInput suggestions={null} />);

    fireEvent.click(getByTestId('search-input'));

    expect(getByTestId('search-input')).toHaveClass(
      'search-content__input--active'
    );
  });
});
