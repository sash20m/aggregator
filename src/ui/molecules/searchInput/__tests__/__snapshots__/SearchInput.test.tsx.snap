// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchInput should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="search-content"
  >
    <div
      class="search-content__input-wrapper"
    >
      <input
        class="search-content__input"
        data-testid="search-input"
        placeholder="Search from 225,195 companies"
        type="text"
        value=""
      />
      <img
        alt=""
        class="search-content__search-magnifier"
        src="../search.png"
      />
    </div>
    <div
      class="search-content__options"
    >
      <p
        class="search-content__options__item"
      >
        Search In
      </p>
      <p
        class="search-content__options__item--selected"
      >
        Companies
      </p>
      <p
        class="search-content__options__item"
      >
        Persons
      </p>
    </div>
  </div>
</DocumentFragment>
`;
